lib/rack/mock_request.rb

```diff
@@ -139,14 +139,18 @@ def self.env_for(uri = "", opts = {})
         end
       end
 
-      if String === opts[:input]
-        rack_input = StringIO.new(opts[:input])
+      input = opts[:input]
+      if String === input
+        rack_input = StringIO.new(input)
+        rack_input.set_encoding(Encoding::BINARY)
       else
-        rack_input = opts[:input]
+        if input.respond_to?(:encoding) && input.encoding != Encoding::BINARY
+          raise ArgumentError, "input must be binary"
```

Comment:
Should we raise an exception, or just emit a warning and still call `set_encoding`?  That way we're backwards compatible and folks can fix their code.

Reply:
People often ignore warnings, and we won't release this until Rack 3.1 so I personally vote for error.

However, I don't have a strong opinion one way or the other. I think the amount of external test code failing because of this will be fairly small.

That being said, it is a breaking change. A warning wouldn't be. We'd still need to write:

```
warn "input encoding not binary", uplevel: 1
if input.respond_to?(:set_encoding)
  input.set_encoding(Encoding::BINARY)
else
  raise ArgumentError, "could not coerce input to binary encoding"
end
```